bitcoin/script/miniscript.h

```diff
@@ -726,167 +748,165 @@ struct Node {
+                case NodeType::WRAP_N:
+                    return std::move(subres[0]);
+                case NodeType::WRAP_D: {
+                    auto &x = subres[0];
+                    return InputResult(ZERO, x.sat + ONE);
+                }
+                case NodeType::WRAP_J: {
+                    auto &x = subres[0];
+                    // If a dissatisfaction with a nonzero top stack element exists, an alternative dissatisfaction exists.
+                    // As the dissatisfaction logic currently doesn't keep track of this nonzeroness property, and thus even
+                    // if a dissatisfaction with a top zero element is found, we don't know whether another one with a
+                    // nonzero top stack element exists. Make the conservative assumption that whenever the subexpression is weakly
+                    // dissatisfiable, this alternative dissatisfaction exists and leads to malleability.
+                    return InputResult(InputStack(ZERO).Malleable(x.nsat.available != Availability::NO && !x.nsat.has_sig), std::move(x.sat));
+                }
+                case NodeType::WRAP_V: {
+                    auto &x = subres[0];
+                    return InputResult(INVALID, std::move(x.sat));
+                }
+                case NodeType::JUST_0: return InputResult(EMPTY, INVALID);
+                case NodeType::JUST_1: return InputResult(INVALID, EMPTY);
             }
-            case NodeType::WRAP_V: {
-                auto x = subs[0]->ProduceInput(ctx, nonmal);
-                return InputResult(INVALID, x.sat);
+            assert(false);
+            return InputResult(INVALID, INVALID);
+        };
+
+        auto tester = [&helper, nonmal](const Node& node, Span<InputResult> subres) -> InputResult {
+            auto ret = helper(node, subres);
+            // Do a consistency check between the satisfaction code and the type checker
+            // (the actual satisfaction code in ProduceInputHelper does not use GetType)
+            if (node.GetType() << "z"_mst && ret.nsat.available != Availability::NO) assert(ret.nsat.stack.size() == 0);
+            if (node.GetType() << "z"_mst && ret.sat.available != Availability::NO) assert(ret.sat.stack.size() == 0);
+            if (node.GetType() << "o"_mst && ret.nsat.available != Availability::NO) assert(ret.nsat.stack.size() == 1);
+            if (node.GetType() << "o"_mst && ret.sat.available != Availability::NO) assert(ret.sat.stack.size() == 1);
+            if (node.GetType() << "n"_mst && ret.sat.available != Availability::NO) assert(ret.sat.stack.back().size() != 0);
+            if (node.GetType() << "d"_mst) assert(ret.nsat.available != Availability::NO);
+            if (node.GetType() << "f"_mst && ret.nsat.available != Availability::NO) assert(ret.nsat.has_sig);
+            if (node.GetType() << "s"_mst && ret.sat.available != Availability::NO) assert(ret.sat.has_sig);
+            if (nonmal) {
```

Comment:
With the above rationale, we should only assert things about nonmal, if the assumption that the top level is "s" is satisfied.

Reply:
I don't think that's necessary, actually. The nonmalleable satisfier for internal nodes of the tree should always have these properties, even if no top-level non-malleable solution exists.